### PR TITLE
Added link latencies to the IPv8 simulator

### PR DESCRIPTION
**Pull request by devos50**
This PR:

 - Adds support for link latencies in the simulator. Specifically, we modified the `SimulationEndpoint` class to achieve this.

### CODE REVIEW

**Comment by devos50** [comment]
validate

File: simulation/simulation_endpoint.py [hunk 1]
```diff
@@ -11,3 +13,16 @@ class SimulationEndpoint(AutoMockEndpoint):
...
+        Return the link latency to a particular destination address, in seconds.
+        """
+        return 0
+
+    async def actual_send(self, socket_address, packet):
+        await sleep(self.get_link_latency(socket_address))
+        super().send(socket_address, packet)
+
+    def send(self, socket_address, packet):
+        ensure_future(self.actual_send(socket_address, packet))
```
Comment (qstokkink) [inline simulation/simulation_endpoint.py:28]: It seems like `actual_send` will not be overwritten in subclasses. If that is the case, please use this instead:

```suggestion
    def send(self, socket_address, packet):
        get_event_loop().call_later(self.get_link_latency(socket_address), super().send,
                                    socket_address, packet)
```

File: simulation/simulation_endpoint.py [hunk 2]
```diff
@@ -11,3 +13,16 @@ class SimulationEndpoint(AutoMockEndpoint):
     async def open(self):
         self._open = True
         return succeed(None)
+
+    def get_link_latency(self, to_address):
+        """
+        Return the link latency to a particular destination address, in seconds.
+        """
+        return 0
```
Comment (qstokkink) [inline simulation/simulation_endpoint.py:21]: This is dead code. Please consider adding something like `self.latencies = defaultdict(int)` and changing this method to return `self.latencies[to_address]`.

**Review by qstokkink — CHANGES_REQUESTED** [review]

File: simulation/simulation_endpoint.py (hunk 1 above)
Comment (devos50) [reply simulation/simulation_endpoint.py:28]: That's a more clean solution indeed, thanks for the suggestion! 👍

**Comment by devos50** [comment]
validate

**Review by qstokkink — APPROVED** [review]
Perfect 👍  thanks!

**Pull request merged**